`ct/l1`: add `compaction_worker` and `worker_manager` [CT - Compaction #4]

WillemKauf opened the pull request:
Getting into some of the fun stuff!

The `compaction_worker` is a per-shard actor that simply invokes the traditional pattern of

```cpp
auto src = std::make_unique<compaction_source>();
auto sink = std::make_unique<compaction_sink>();
auto reducer = compaction::sliding_window_reducer(std::move(src), std::move(sink));
auto compact_fut = co_await ss::coroutine::as_future(std::move(reducer).run());
```

while managing some state and allowing for some control over compaction jobs, either by a graceful cancellation request (e.g. due to pre-emption) or a hard stop request (e.g. during shutdown).

Above the `compaction_worker` sits the `worker_manager`, which manages a `ss::sharded<>` pool of these workers which pull jobs from a priority queue, managed by the `compaction_scheduler` (still a black box to ye reviewers, but its coming eventually) to available workers (e.g. one compaction job may be ran per shard/worker).

The executor does not make informed scheduling decisions, but does keep state around inflight compaction jobs and the workers themselves. 

Soon, it will be possible to (through the V2 AdminAPI) customize which shards/workers the `executor` will dispatch compaction jobs to. I.e, users will be able to customize how many CPU resources to dedicate to cloud topics compaction (for example, I want half my shards dedicated to compaction, or I only want shard 0 dedicated to compaction, or shards 0 and 15, etc...).

How cool is that!

Note that the dependencies on `redpanda` broker state still remain minimal.

## Backports Required

- [X] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes

* none

WillemKauf committed:
`ct/l1`: add comments and additional state to `compaction_job_state`

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--138cc566-11b3-4cc1-a92e-18ec72b69047-->
<!--TEST RESULTS-->
<!--4a1d2188-2144-4475-8355-c5409f1f619f-->
<!--TEST RESULTS-->
<!--58fae34e-728d-48f2-b3b8-4e4c1454eb78-->
<!--TEST RESULTS-->
<!--8a363fd8-b9af-4866-a416-411e0dbf521a-->
<!--TEST RESULTS-->
<!--081defa9-8023-4149-af0c-d7a53c1023cf-->
<!--TEST RESULTS-->
<!--04bc255d-8606-4c46-8ad0-8ba806274d10-->
<!--TEST RESULTS-->
<!--897819a6-713b-459c-89a6-00923da7c729-->
<!--TEST RESULTS-->
<!--9d7d319c-e768-41eb-a012-4bed40769fad-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/72567">build#72567</a></summary>


| test_class                           | test_method                             | test_arguments                                                                    | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 | test_history                                                                                                                                                                      |
|:-------------------------------------|:----------------------------------------|:----------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ShadowLinkConsumeGroupsMirroringTest | test_continuous_group_sync              | {"source_cluster_spec": {"cluster_type": "redpanda"}, "with_failures": false}     | integration | https://buildkite.com/redpanda/redpanda/builds/72567#01995eda-1335-4596-aa93-5bc027f948c0 | FLAKY         | 20/21    | upstream reliability is '93.75'. current run reliability is '95.23809523809523'. drift is -1.4881 and the allowed drift is set to 50. The test should PASS             | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkConsumeGroupsMirroringTest&test_method=test_continuous_group_sync     |
| DatalakeE2ETests                     | test_avro_schema                        | {"catalog_type": "rest_hadoop", "cloud_storage_type": 1, "query_engine": "trino"} | integration | https://buildkite.com/redpanda/redpanda/builds/72567#01995eda-1337-4e49-8891-140e833dea10 | FLAKY         | 20/21    | upstream reliability is '99.48186528497409'. current run reliability is '95.23809523809523'. drift is 4.24377 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DatalakeE2ETests&test_method=test_avro_schema                                   |
| IdempotencyWriteCachingTest          | test_idempotent_producers_write_caching | null                                                                              | integration | https://buildkite.com/redpanda/redpanda/builds/72567#01995ed3-d822-4c1e-b78c-446b8b49a122 | FLAKY         | 20/21    | upstream reliability is '99.48051948051948'. current run reliability is '95.23809523809523'. drift is 4.24242 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=IdempotencyWriteCachingTest&test_method=test_idempotent_producers_write_caching |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/72742">build#72742</a></summary>


| test_class                          | test_method                                       | test_arguments                                                                                                                           | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                   | test_history                                                                                                                                                                         |
|:------------------------------------|:--------------------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ClusterRateQuotaTest                | test_client_group_consume_rate_throttle_mechanism | null                                                                                                                                     | integration | https://buildkite.com/redpanda/redpanda/builds/72742#01997750-40fa-4a61-8cdb-9f087372bcc4 | FLAKY         | 16/21    | upstream reliability is '85.86956521739131'. current run reliability is '76.19047619047619'. drift is 9.67909 and the allowed drift is set to 50. The test should PASS   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ClusterRateQuotaTest&test_method=test_client_group_consume_rate_throttle_mechanism |
| ClusterRateQuotaTest                | test_client_response_throttle_mechanism           | null                                                                                                                                     | integration | https://buildkite.com/redpanda/redpanda/builds/72742#01997750-40f4-43e1-8f06-a4994321d783 | FLAKY         | 18/21    | upstream reliability is '93.61179361179362'. current run reliability is '85.71428571428571'. drift is 7.89751 and the allowed drift is set to 50. The test should PASS   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ClusterRateQuotaTest&test_method=test_client_response_throttle_mechanism           |
| ClusterRateQuotaTest                | test_client_response_throttle_mechanism           | null                                                                                                                                     | integration | https://buildkite.com/redpanda/redpanda/builds/72742#01997758-0412-495b-8beb-c97b69f0e692 | FLAKY         | 14/21    | upstream reliability is '85.01144164759725'. current run reliability is '66.66666666666666'. drift is 18.34477 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ClusterRateQuotaTest&test_method=test_client_response_throttle_mechanism           |
| SchemaEvolutionE2ETests             | test_old_schema_writer                            | {"catalog_type": "rest_hadoop", "cloud_storage_type": 1, "produce_mode": "avro", "query_engine": "trino", "test_case": "promote_column"} | integration | https://buildkite.com/redpanda/redpanda/builds/72742#01997750-40f7-454c-8c9f-4deeab06c64d | FLAKY         | 20/21    | upstream reliability is '99.10313901345292'. current run reliability is '95.23809523809523'. drift is 3.86504 and the allowed drift is set to 50. The test should PASS   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=SchemaEvolutionE2ETests&test_method=test_old_schema_writer                         |
| DisablingPartitionsTest             | test_disable                                      | null                                                                                                                                     | integration | https://buildkite.com/redpanda/redpanda/builds/72742#01997750-40f9-4164-90c7-bf7a17c20e0c | FLAKY         | 13/21    | upstream reliability is '86.28930817610063'. current run reliability is '61.904761904761905'. drift is 24.38455 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DisablingPartitionsTest&test_method=test_disable                                   |
| WriteCachingFailureInjectionE2ETest | test_crash_all                                    | {"use_transactions": false}                                                                                                              | integration | https://buildkite.com/redpanda/redpanda/builds/72742#01997750-40f2-442d-94a6-cad96b9c00fb | FLAKY         | 18/21    | upstream reliability is '98.96103896103897'. current run reliability is '85.71428571428571'. drift is 13.24675 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=WriteCachingFailureInjectionE2ETest&test_method=test_crash_all                     |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/72767">build#72767</a></summary>


| test_class                          | test_method                                       | test_arguments                  | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                   | test_history                                                                                                                                                                         |
|:------------------------------------|:--------------------------------------------------|:--------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| AuditLogTestKafkaApi                | test_no_auth_enabled                              | {"audit_transport_mode": "rpc"} | integration | https://buildkite.com/redpanda/redpanda/builds/72767#019977e3-3a18-49a2-aa75-7eaf63aa12ee | FLAKY         | 19/21    | upstream reliability is '100.0'. current run reliability is '90.47619047619048'. drift is 9.52381 and the allowed drift is set to 50. The test should PASS               | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=AuditLogTestKafkaApi&test_method=test_no_auth_enabled                              |
| EndToEndCloudTopicsTest             | test_write                                        | null                            | integration | https://buildkite.com/redpanda/redpanda/builds/72767#019977e3-3a1a-4026-89e7-10d98a78c27c | FLAKY         | 17/21    | upstream reliability is '89.15094339622641'. current run reliability is '80.95238095238095'. drift is 8.19856 and the allowed drift is set to 50. The test should PASS   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=EndToEndCloudTopicsTest&test_method=test_write                                     |
| EndToEndCloudTopicsTxTest           | test_write                                        | null                            | integration | https://buildkite.com/redpanda/redpanda/builds/72767#019977e3-3a1b-4704-8e28-9c0887842988 | FLAKY         | 14/21    | upstream reliability is '85.22571819425444'. current run reliability is '66.66666666666666'. drift is 18.55905 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=EndToEndCloudTopicsTxTest&test_method=test_write                                   |
| EndToEndCloudTopicsTxTest           | test_write                                        | null                            | integration | https://buildkite.com/redpanda/redpanda/builds/72767#019977e9-5c74-4419-b5bc-f544d623a693 | FLAKY         | 20/21    | upstream reliability is '85.22571819425444'. current run reliability is '95.23809523809523'. drift is -10.01238 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=EndToEndCloudTopicsTxTest&test_method=test_write                                   |
| ClusterRateQuotaTest                | test_client_group_consume_rate_throttle_mechanism | null                            | integration | https://buildkite.com/redpanda/redpanda/builds/72767#019977e9-5c74-4419-b5bc-f544d623a693 | FLAKY         | 13/21    | upstream reliability is '85.98382749326146'. current run reliability is '61.904761904761905'. drift is 24.07907 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ClusterRateQuotaTest&test_method=test_client_group_consume_rate_throttle_mechanism |
| ClusterRateQuotaTest                | test_client_group_produce_rate_throttle_mechanism | null                            | integration | https://buildkite.com/redpanda/redpanda/builds/72767#019977e9-5c65-4657-b942-57465890c00a | FLAKY         | 17/21    | upstream reliability is '88.4437596302003'. current run reliability is '80.95238095238095'. drift is 7.49138 and the allowed drift is set to 50. The test should PASS    | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ClusterRateQuotaTest&test_method=test_client_group_produce_rate_throttle_mechanism |
| ClusterRateQuotaTest                | test_client_response_throttle_mechanism           | null                            | integration | https://buildkite.com/redpanda/redpanda/builds/72767#019977e9-5c69-4e10-a7ee-f5efa9c4a54f | FLAKY         | 17/21    | upstream reliability is '84.61538461538461'. current run reliability is '80.95238095238095'. drift is 3.663 and the allowed drift is set to 50. The test should PASS     | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ClusterRateQuotaTest&test_method=test_client_response_throttle_mechanism           |
| DisablingPartitionsTest             | test_disable                                      | null                            | integration | https://buildkite.com/redpanda/redpanda/builds/72767#019977e9-5c70-4170-a36f-4aa8d6636972 | FLAKY         | 12/21    | upstream reliability is '86.09112709832134'. current run reliability is '57.14285714285714'. drift is 28.94827 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DisablingPartitionsTest&test_method=test_disable                                   |
| WriteCachingFailureInjectionE2ETest | test_crash_all                                    | {"use_transactions": false}     | integration | https://buildkite.com/redpanda/redpanda/builds/72767#019977e3-3a13-4984-b2e1-9012a3ea578a | FLAKY         | 19/21    | upstream reliability is '98.05825242718447'. current run reliability is '90.47619047619048'. drift is 7.58206 and the allowed drift is set to 50. The test should PASS   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=WriteCachingFailureInjectionE2ETest&test_method=test_crash_all                     |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/73282">build#73282</a></summary>


| test_class               | test_method          | test_arguments                                                                                                                         | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 | test_history                                                                                                                                                |
|:-------------------------|:---------------------|:---------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------------------------------------------------------------------------------------------------------------------------------------------------------|
| RandomNodeOperationsTest | test_node_operations | {"cloud_storage_type": 1, "compaction_mode": "sliding_window", "enable_failures": true, "mixed_versions": true, "with_iceberg": false} | integration | https://buildkite.com/redpanda/redpanda/builds/73282#01999c73-4bee-47e6-a319-97a0fdee709b | FLAKY         | 20/21    | upstream reliability is '99.16666666666667'. current run reliability is '95.23809523809523'. drift is 3.92857 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RandomNodeOperationsTest&test_method=test_node_operations |
| TieredStorageTest        | test_tiered_storage  | {"cloud_storage_type_and_url_style": [2, "virtual_host"], "test_case": {"name": "(TS_Read == True, TS_Timequery == True)"}}            | integration | https://buildkite.com/redpanda/redpanda/builds/73282#01999c73-4bfd-4398-9fee-6c1c6abff10e | FAIL          | 0/1      |                                                                                                                                                                        | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=TieredStorageTest&test_method=test_tiered_storage         |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/73379">build#73379</a></summary>


| test_class                       | test_method                           | test_arguments                                         | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  | test_history                                                                                                                                                                   |
|:---------------------------------|:--------------------------------------|:-------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| DataMigrationsApiTest            | test_creating_and_listing_migrations  | null                                                   | integration | https://buildkite.com/redpanda/redpanda/builds/73379#0199a159-803a-454d-b9bd-535abc9899c4 | FLAKY         | 20/21    | upstream reliability is '99.79633401221996'. current run reliability is '95.23809523809523'. drift is 4.55824 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DataMigrationsApiTest&test_method=test_creating_and_listing_migrations       |
| DatalakeClusterRestoreTest       | test_slow_tiered_storage_dupe_records | {"catalog_type": "rest_jdbc", "cloud_storage_type": 1} | integration | https://buildkite.com/redpanda/redpanda/builds/73379#0199a1a5-1257-426b-b333-83678bcd7100 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS               | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DatalakeClusterRestoreTest&test_method=test_slow_tiered_storage_dupe_records |
| OffsetRetentionTest              | test_offset_expiration                | null                                                   | integration | https://buildkite.com/redpanda/redpanda/builds/73379#0199a15f-e361-4601-a9d3-af19b9ed3596 | FAIL          | 0/1      |                                                                                                                                                                         | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=OffsetRetentionTest&test_method=test_offset_expiration                       |
| OffsetRetentionTest              | test_offset_expiration                | null                                                   | integration | https://buildkite.com/redpanda/redpanda/builds/73379#0199a1a5-1257-426b-b333-83678bcd7100 | FAIL          | 0/1      | The test has failed across all retries                                                                                                                                  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=OffsetRetentionTest&test_method=test_offset_expiration                       |
| TimeQueryTest                    | test_timequery_with_local_gc          | null                                                   | integration | https://buildkite.com/redpanda/redpanda/builds/73379#0199a15f-e361-4c6c-8242-d73b423e24d5 | FLAKY         | 19/21    | upstream reliability is '92.71255060728745'. current run reliability is '90.47619047619048'. drift is 2.23636 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=TimeQueryTest&test_method=test_timequery_with_local_gc                       |
| WriteCachingFailureInjectionTest | test_unavoidable_data_loss            | null                                                   | integration | https://buildkite.com/redpanda/redpanda/builds/73379#0199a159-803a-454d-b9bd-535abc9899c4 | FLAKY         | 17/21    | upstream reliability is '94.30051813471503'. current run reliability is '80.95238095238095'. drift is 13.34814 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=WriteCachingFailureInjectionTest&test_method=test_unavoidable_data_loss      |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/73558">build#73558</a></summary>


| test_class                          | test_method                  | test_arguments              | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 | test_history                                                                                                                                                     |
|:------------------------------------|:-----------------------------|:----------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------|
| PartitionBalancerTest               | test_rack_awareness          | null                        | integration | https://buildkite.com/redpanda/redpanda/builds/73558#0199ab52-3380-45c8-987e-a3091650ed38 | FAIL          | 0/1      |                                                                                                                                                                        | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=PartitionBalancerTest&test_method=test_rack_awareness          |
| TimeQueryTest                       | test_timequery_with_local_gc | null                        | integration | https://buildkite.com/redpanda/redpanda/builds/73558#0199ab52-337f-450b-ab06-7d50d6d5a03b | FLAKY         | 19/21    | upstream reliability is '92.08633093525181'. current run reliability is '90.47619047619048'. drift is 1.61014 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=TimeQueryTest&test_method=test_timequery_with_local_gc         |
| WriteCachingFailureInjectionE2ETest | test_crash_all               | {"use_transactions": false} | integration | https://buildkite.com/redpanda/redpanda/builds/73558#0199ab52-337a-4b0a-82f4-b7c262388780 | FLAKY         | 16/21    | upstream reliability is '84.59016393442623'. current run reliability is '76.19047619047619'. drift is 8.39969 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=WriteCachingFailureInjectionE2ETest&test_method=test_crash_all |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/73609">build#73609</a></summary>


| test_class        | test_method         | test_arguments                                                                                                              | test_kind   | job_url                                                                                   | test_status   | passed   | reason   | test_history                                                                                                                                        |
|:------------------|:--------------------|:----------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------|
| TieredStorageTest | test_tiered_storage | {"cloud_storage_type_and_url_style": [1, "virtual_host"], "test_case": {"name": "(TS_Read == True, TS_Timequery == True)"}} | integration | https://buildkite.com/redpanda/redpanda/builds/73609#0199b1be-b54d-4027-9375-dfc148db4be9 | FAIL          | 0/1      |          | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=TieredStorageTest&test_method=test_tiered_storage |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/73663">build#73663</a></summary>


| test_class             | test_method                 | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 | test_history                                                                                                                                                     |
|:-----------------------|:----------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------|
| RapidTopicRecreateTest | test_topic_rapid_recreation | null             | integration | https://buildkite.com/redpanda/redpanda/builds/73663#0199bbd4-4cb4-4a6e-8f4b-31744dd662bb | FLAKY         | 20/21    | upstream reliability is '99.29411764705883'. current run reliability is '95.23809523809523'. drift is 4.05602 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RapidTopicRecreateTest&test_method=test_topic_rapid_recreation |
</details>
<!--/body-->

rockwotj commented on src/v/cloud_topics/level_one/compaction/source.h:
```diff
@@ -18,7 +18,19 @@
+    idle,
+    // A compaction job is currently inflight.
+    running,
+    // A graceful stop has been requested of an inflight compaction job.
+    // The user should try to commit as much useful data as possible while still
+    // shutting down in a prompt manner.
+    cancelled,
+    // A forceful stop has been requested of an inflight compaction job.
+    // The user should abandon any work and shutdown immediately.
+    stopped
```
`cancelled` makes me think of an abrupt stop... Struggling with better names, in RPCN we call this `soft_stop` and `hard_stop`.

WillemKauf replied:
Those are better names, thanks.

WillemKauf replied:
Changed

dotnwat commented on src/v/cloud_topics/level_one/compaction/executor.h:
```diff
@@ -0,0 +1,236 @@
+    // Waits for an available worker from the pool and then issues a
+    // backgrounded compaction job for the provided `log` on that worker's
+    // shard.
+    ss::future<> compact_log(log_info_and_meta);
```
what's the purpose of waiting for a worker to become available if the caller isn't going to wait for the entire compaction job to run? it would seem that the caller only cares that the job is enqueued.

WillemKauf replied:
Fair, maybe it would make more sense if this future only resolved when the entire compaction job was complete.

dotnwat commented on src/v/cloud_topics/level_one/compaction/executor.h:
```diff
@@ -0,0 +1,236 @@
+    // compaction jobs to be ran. This function is ideally used when e.g. a
+    // partition is removed or the `cleanup.policy` for a topic is changed and a
+    // single compaction job must be stopped.
+    ss::future<> request_stop_compaction(model::ntp);
```
other parts of L1 seem to be using partition IDs now. should we do that here?

WillemKauf replied:
Yeah, I figured I would go back eventually when the `ntp -> tidp` migration was more settled, but I'm happy to change this stuff now.

WillemKauf replied:
Altered to use `tidp` everywhere

dotnwat commented on src/v/cloud_topics/level_one/compaction/executor.h:
```diff
@@ -0,0 +1,236 @@
+        // compaction job on this worker. If not set, the worker is idle.
+        std::optional<model::ntp> ntp{std::nullopt};
+
+        intrusive_list_hook link;
```
- IIUC there are smp::count workers? seems like we could simplify a bit by not dealing with an intrusive list?

WillemKauf replied:
The use of the `intrusive_list` is mostly to support

> Soon, it will be possible to (through the V2 AdminAPI) customize which shards/workers the executor will dispatch compaction jobs to. I.e, users will be able to customize how many CPU resources to dedicate to cloud topics compaction (for example, I want half my shards dedicated to compaction, or I only want shard 0 dedicated to compaction, or shards 0 and 15, etc...).

The worker pool is ideally a FIFO like structure that also supports removal of an arbitrary element per the above, which I felt the `intrusive_list` was a good fit for.

If this seems overly complicated or if you have any other suggestions please let me know.

rockwotj replied:
Seems like you need a set of these worker_meta objects and a number of indexes on them right? By ntp, a fifo of "last ran a job on" and a way to look one up by shard/id?

Honestly since cores are bound to at most be 256 or something in seastar, using boost multi index for all this seems reasonable if that works (no issues with oversized allocs), otherwise you could have a absl::FixedArray<std::optional<worker_meta>> and everything else is the index/shard id. I am not sure how much more of this is evolving to give more guidance on which is the right way to go.

WillemKauf replied:
> I am not sure how much more of this is evolving to give more guidance on which is the right way to go.

Ideally the worker & executor won't evolve much past this point. I just want the design to support:

* Enabling/pausing `compaction_worker`s on arbitrary shards
* An implementation that distributes a number of compaction jobs at a time amongst the worker pool and resolves a future when all are complete (or stopped).

Putting either a pull or a push based model here feels like it leads to quirks/complexity in the implementation, but if you think one model or the other, or perhaps different data structure(s) would make the implementation more straight forward I'm happy to rework it.

How I see it:

Pull-based:
- Pro: Less book-keeping in the executor, it doesn't have to manage much beyond having a queue of jobs to distribute and a container of inflight compaction jobs `_inflight_ntp_to_shard`. Pausing workers is easier because we aren't managing a pool of available workers.
- Con: More cross-shard communication (`worker` queries `executor` on shard 0 for a entry in its job queue, steals it, performs the work, then likely sets a `promise`'s `future<>` again on shard 0 to indicate the job is done to the higher level waiter)

Push-based:
- Pro: Pushing and waiting on work to be done seems more straight forward, the `executor` distributes jobs to workers when they become available and can wait for all compaction futures to resolve.
- Con: More book-keeping/data structures in the `executor` to support enabling & pausing of `worker`s

rockwotj replied:
> How I see it:

The way I read it seems much simpler to do pull based. We're already pulling from the metastore, so unless the bookkeeping is needed to unlock scheduling decisions it seems like the way to go. I think the pull model is also more amendable to add extra dedicated compute for compaction.

WillemKauf replied:
Changed to a pull based model.

dotnwat commented on src/v/cloud_topics/level_one/compaction/executor.cc:
```diff
@@ -0,0 +1,195 @@
+            // `_avail_workers` pool now and alert a waiter.
+            auto is_inflight = meta.ntp.has_value();
+            if (!is_inflight) {
+                _avail_workers.push_back(meta);
```
i'm wondering if the executor might be a bit more complex than necessary because it's tracking all the per-worker state? would it be sufficient to model a worker as an entity that starts on each core and only knows how to ask for work (or be told what to do)?

so like a model for a worker might be:

```
loop:
  work = ask executor for work
  do work
```

then what is the job of the executor? maybe it is just maintaining a single data structure which contains the current job queue for each worker, which it can change at any time.

if it wants to reschedule something on a worker, it can first change the queue, then reach out and stop that worker and then do nothing else: the worker will be expected to ask for new work when its done cleaning up.

WillemKauf replied:
> to ask for work (or be told what to do)?

Yeah, the worker could be a push or pull actor. If it were a pull based model it might make the pausing/disabling of a worker much simpler, and as a result, the entire implementation as well...

dotnwat reviewed (commented):
this is looking pretty good. i didn't do a thorough review, but added some notes about the executor/worker model. it seems like there are quite a few states being tracked, and i'm wondering if perhaps the model can be simplified a bit? maybe not; it's just a quick reaction.

rockwotj commented on src/v/cloud_topics/level_one/compaction/executor.h:
```diff
@@ -0,0 +1,237 @@
+// TODO: Hook this up to the AdminAPI to allow for users to customize which
+// shards have active `compaction_worker`s, and persist that information in e.g.
+// the kvstore.
+class compaction_executor {
```
This is a a singleton that lives on shard0?

WillemKauf replied:
Yes it is, I will add that in a comment.

WillemKauf replied:
Added

WillemKauf commented:
Force push to rebase to `upstream/dev`

dotnwat commented on src/v/cloud_topics/level_one/compaction/worker.h:
```diff
@@ -0,0 +1,94 @@
+    compaction_worker(io*, metastore*, compaction_committer*);
+
+    // Requests a compaction of the provided CTP and its `compaction_offsets`
+    // as obtained from the `metastore`.
```
I think that this method (and perhaps by extention the compaction_worker type) could use some additional explanation. The idea of a "worker" running "jobs" makes sense, but then we have this interface to "request a compaction", but from what I can tell it builds a reducer and does compaction work in the foreground. That's not quite what I would expect from an abstraction for requesting a worker to manage a job.

Specifically, I'm not clear what the difference between a worker and a job is, primarily because the worker seems to have methods (e.g. initialize_map) that I would expect a "job" to own. That is, why is a worker doing more than merely managing execution of a job?

WillemKauf replied:
Please let me know if the same concerns exist in the reworked pull based implementation, or if this is still far off of your mental model for what this worker pool architecture should look/feel like.

dotnwat commented on src/v/cloud_topics/level_one/compaction/executor.cc:
```diff
@@ -0,0 +1,202 @@
+    _as.request_abort();
+    _cvar.broken();
+    co_await _workers.invoke_on_all(
+      [](compaction_worker& worker) { worker.stop_worker(); });
```
maybe it would be useful to have some different terminology for these different scenarios. like request_stop_workers seems like it does _more_ than `stop` but the later should do _more_ since it's shutting down the entire executor?

stop for reference:

```
ss::future<> compaction_executor::stop() { co_await _workers.stop(); }
```

WillemKauf replied:
Thanks, reworked this a little bit. The stopping of workers occurs in `worker_manager::stop()` before they are destructed now.

dotnwat commented on src/v/cloud_topics/level_one/compaction/executor.cc:
```diff
@@ -0,0 +1,202 @@
+    auto it = _worker_meta.find(worker);
+
+    if (it == _worker_meta.end()) {
+        throw std::runtime_error("This should never happen.");
```
you can vassert if that makes sense, or return std::expected as the preferred method for communicating errors.

dotnwat commented on src/v/cloud_topics/level_one/compaction/executor.cc:
```diff
@@ -0,0 +1,202 @@
+
+compaction_executor::worker_meta&
+compaction_executor::get_worker_meta(worker_shard worker) {
+    auto it = _worker_meta.find(worker);
```
it looks like `_worker_meta` should just be a std::array[smp::count] ?

dotnwat commented on src/v/cloud_topics/level_one/compaction/executor.h:
```diff
@@ -0,0 +1,239 @@
+    using worker_set = chunked_hash_set<
+      worker_meta_ptr,
+      worker_meta_ptr_hash,
+      worker_meta_ptr_eq>;
```
afaict this can be a `std::vector<shard>`?

rockwotj replied:
yeah doesn't need to be chunked

dotnwat commented on src/v/cloud_topics/level_one/compaction/executor.cc:
```diff
@@ -0,0 +1,202 @@
+    }
+
+    meta.current_state = new_state;
+    auto is_linked = meta.link.is_linked();
```
is it true that this could also be named `is_available`? if there is only 1 worker per core, i'm wondering if things would be simpler if all the state were in the worker meta object (e.g. `is_available: boolean`). just a thought. then you can just look for a worker in the array/vector that is available.

dotnwat reviewed (commented):
im still a bit confused about what a job is. it doesn't seem like a job is an explicit thing that is tracked?

rockwotj reviewed (commented):
I don't really have much to add to Noah's comments.

WillemKauf committed:
`l1/compaction`: add `compaction_info_and_timestamp` to `meta`

To track `compaction_info` sampled from the `metastore` as well as
the timestamp at when it was sampled.

Also adds the `compaction_log_queue` typedef, which is a priority queue
holding pointers to `log_compaction_meta`s which require compaction.
The idea here is that a `log_compaction_meta*` with `info_and_ts` set
and requiring compaction should be pushed into a `priority_queue` held
by a scheduler/manager in order to allow `compaction_worker`s to pull
jobs from the global queue and complete them on their shard.

WillemKauf committed:
`l1/compaction`: add `compaction_worker`

The object responsible for creating/managing a compaction `source/sink/reducer`
for `l1` compaction.

This will eventually be ran per shard within a `worker_manager`, which will
be responsible for managing the state and inflight compaction jobs ran on a
`compaction_worker`.

WillemKauf committed:
`l1/compaction`: add `worker_manager`

Owns a sharded pool of workers on which compaction jobs are ran.
The idea is to one day be able to set which shards have active compaction
workers through the Admin API, allowing users to fully customize how many
CPU resources are allocated to compaction of cloud topics.

WillemKauf committed:
`l1/compaction`: add `worker_manager_test`

WillemKauf commented:
Force push to rework this PR into a pull based model for the `compaction_worker`. `compaction_executor` is no longer an `executor`, really. It has been renamed to `worker_manager`.

A lot of the review comments are now a bit outdated, but I appreciate all the feedback thus far and would like to hear what people think of the new design. Very open to external discussions around any of the implementation here.

WillemKauf commented:
Force push to rebase to `upstream/dev`.

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/73379-->
<!--1957083c-322c-4287-a826-b0a2f7a17f94-->
<!--8cf69a4b-e08a-41ad-9f45-8bb1525e68a8-->
<!--73ca8f16-f9d7-4994-829e-c7e027311bcd-->
<!--title-->
## Retry command for [Build#73379](https://buildkite.com/redpanda/redpanda/builds/73379)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/offset_retention_test.py::OffsetRetentionTest.test_offset_expiration
```
<!--/body-->

rockwotj commented on src/v/cloud_topics/level_one/compaction/meta.h:
```diff
@@ -94,4 +107,11 @@ struct object_output_t {
+  = std::function<bool(const log_compaction_meta*, const log_compaction_meta*)>;
+using log_compaction_queue = std::priority_queue<
+  log_compaction_meta*,
+  chunked_vector<log_compaction_meta*>,
```
consider using `ss::lw_shared_ptr`, this will prevent issues from someone taking a pointer to a member in `chunked_vector<log_compaction_meta>` or other containers that can move values around, as well as other lifetime guarantees. The main reason I bring this up is that the queue type being defined here makes me thing it's a non local abstraction, which means reasoning about the pointer lifetimes might be difficult.

WillemKauf replied:
I agree, this usage of a raw pointer is unsafe, I may have drawn too much inspiration from `log_manager`.

The `log_compaction_queue` may hold pointers to `log_compaction_meta`s that have long been extracted from the owning container and destructed due to e.g. partition removal while the compaction job has been sitting waiting in the queue, and there is no guarantee by the time we get to processing the entry from a `log_compaction_queue` that the object isn't destroyed.

Changed this to a `lw_shared_ptr` since lifetimes are hard to reason about, and we can still check `log->link.is_linked()` before holding a gate and compacting to ensure we aren't compacting a removed log.

WillemKauf replied:
This isn't safe either though. The idea was to pop from the queue on shard0:

https://github.com/redpanda-data/redpanda/blob/55b2fa305df7a7fb2c0072ff78e01d0801e94607/src/v/cloud_topics/level_one/compaction/worker.cc#L176-L181

and use the `log_compaction_meta_ptr` on a different shard. Argh.

rockwotj replied:
Oh use foreign_ptr then

rockwotj replied:
https://github.com/scylladb/seastar/blob/baad803b10f36eccfd18ddd204ae1c0e42387928/include/seastar/core/sharded.hh#L943

`ss::foreign_ptr<ss::lw_shared_ptr<...>>`. It's a lot to type but useful to signal cross core pointers are being held

rockwotj commented on src/v/cloud_topics/level_one/compaction/worker.h:
```diff
@@ -0,0 +1,165 @@
+    // This function stops the inflight compaction job but does not affect
+    // the worker state- the worker will continue to accept compaction jobs
+    // after this function is called.
+    void request_hard_stop();
```
the `stop` word here is tripping me up with the `stop` above, since it's really only the currently running compaction job, maybe `interrupt_current_job()` and `terminate_current_job()` to take inspiration from unix signals.

WillemKauf replied:
SGTM!

rockwotj commented on src/v/cloud_topics/level_one/compaction/worker_manager.cc:
```diff
@@ -0,0 +1,100 @@
+
+ss::future<> worker_manager::start() {
+    co_await _workers.start(
+      this,
```
alternatively, if you want to break the circular dependency, I would recommend just passing in a ss::noncopyable_function or something.

WillemKauf replied:
May revisit this in the future

rockwotj reviewed (dismissed):
LGTM, just some naming nits, which I am fine doing in a followup

WillemKauf commented:
>which I am fine doing in a followup

pushed

rockwotj commented on src/v/cloud_topics/level_one/compaction/meta.h:
```diff
@@ -94,4 +102,11 @@ struct object_output_t {
     std::unique_ptr<staging_file> staging_file;
 };
 
+using cmp_t = std::function<bool(
+  const log_compaction_meta_ptr, const log_compaction_meta_ptr)>;
```
nit: consider using `const ptr&` to prevent incrementing and decrementing the ref count for every comparison in a sort.

WillemKauf replied:
Done! thanks

dotnwat reviewed (approved):
thanks. the pull model is much easier to following along. nice work!

dotnwat committed:
Merge branch 'dev' into 4_ct_compact
